File: python/src/bench_sdk/evaluation.py
"""Application checks share the versioned Bench report contract, without network work."""
from __future__ import annotations

import asyncio
import hashlib
import inspect
import json
import re
import threading
import time
from dataclasses import dataclass
from typing import Any, Callable

_ALIASES = {"expected_output": "expectedOutput", "expected_state": "expectedState", "business_outcome": "businessOutcome",
            "required_tools": "requiredTools", "forbidden_tools": "forbiddenTools", "max_tool_calls": "maxToolCalls", "max_model_calls": "maxModelCalls"}


def snapshot(value: Any) -> Any:
    return json.loads(json.dumps(value, allow_nan=False))


def equal(a: Any, b: Any) -> bool:
    if isinstance(a, bool) or isinstance(b, bool):
        return type(a) is type(b) and a == b
    if isinstance(a, dict) and isinstance(b, dict):
        return a.keys() == b.keys() and all(equal(a[k], b[k]) for k in a)
    if isinstance(a, list) and isinstance(b, list):
        return len(a) == len(b) and all(equal(x, y) for x, y in zip(a, b))
    return a == b


@dataclass(frozen=True)
class EvaluationContext:
    case_id: str
    signal: threading.Event
    deadline: float
    turn_index: int | None = None

    @property
    def cancelled(self) -> bool:
        return self.signal.is_set() or time.monotonic() >= self.deadline

    def raise_if_cancelled(self) -> None:
        if self.cancelled:
            raise asyncio.CancelledError()


class Capture:
    def __init__(self):
        self.lock = threading.Lock()
        self.active, self.limited, self.pending = True, False, 0
        self.spans: list[dict] = []

    def start(self):
        with self.lock:
            self.pending += 1

    def finish(self, row):
        with self.lock:
            self.pending -= 1
            if self.active:
                if row is None or len(self.spans) >= 100:
                    self.limited = True
                else:
                    self.spans.append(row)

    def freeze(self):
        with self.lock:
            self.active = False
            return list(self.spans), self.limited, self.pending


def validate(source_revision, context_revision, cases, timeout):
    if not isinstance(source_revision, str) or not re.fullmatch(r"[a-f0-9]{40}", source_revision) or not isinstance(context_revision, str) or not 1 <= len(context_revision) <= 200:
        raise ValueError("Pin a full source revision and a context revision.")
    if not isinstance(cases, list) or not 1 <= len(cases) <= 100 or isinstance(timeout, bool) or not 0 < timeout <= 300:
        raise ValueError("Use 1 to 100 cases and a timeout up to 300 seconds.")
    pinned = snapshot(cases)
    if len(json.dumps(pinned).encode()) > 500000:
        raise ValueError("Case suite exceeds 500 KB.")
    ids = set()
    for case in pinned:
        if not isinstance(case, dict):
            raise ValueError("A case must be an object.")
        for key, alias in _ALIASES.items():
            if key in case:
                if alias in case:
                    raise ValueError("Use one spelling per case field.")
                case[alias] = case.pop(key)
        name = case.get("id")
        if not isinstance(name, str) or not 1 <= len(name) <= 100 or name in ids:
            raise ValueError("Case IDs must be unique and nonempty.")
        ids.add(name)
        if "input" not in case or case.get("split", "regression") not in ("capability", "incident", "regression", "holdout"):
            raise ValueError("Each case needs an input and valid split.")
        if "businessOutcome" in case and (not isinstance(case["businessOutcome"], str) or not 1 <= len(case["businessOutcome"]) <= 2000):
            raise ValueError("Provide a bounded business outcome.")
        count = 0
        for key in ("requiredTools", "forbiddenTools"):
            names = case.get(key, [])
            if not isinstance(names, list) or any(not isinstance(n, str) or not 1 <= len(n) <= 200 for n in names):
                raise ValueError("Invalid tool assertions.")
            count += len(names)
        if count > 90:
            raise ValueError("Use at most 90 tool assertions per case.")
        for key in ("maxToolCalls", "maxModelCalls"):
            if key in case and (type(case[key]) is not int or not 0 <= case[key] <= 10000):
                raise ValueError("Call limits must be nonnegative integers up to 10000.")
    return pinned


async def invoke(fn: Callable, *args):
    # Sync callbacks run off-loop. Cancellation cannot kill a Python thread.
    result = fn(*args) if inspect.iscoroutinefunction(fn) else await asyncio.to_thread(fn, *args)
    return await result if inspect.isawaitable(result) else result


def score(case, output, state, observed, spans, error):
    if not any(not s.get("parent_span_id") and s["kind"] == "AGENT" and s["status"] == "ok" for s in spans):
        error = error or "Application root was not recorded. No complete score."
    if "expectedState" in case and not observed:
        error = error or "Application state was not observed. No complete score."
    checks, findings = [], []
    def check(id, passed, reason):
        checks.append(dict(id=id, passed=passed, reason=reason))
    if "expectedOutput" in case:
        check("expected-output", equal(output, case["expectedOutput"]), "Compare actual output with the expected outcome.")
    if "expectedState" in case and observed:
        check("expected-state", equal(state, case["expectedState"]), "Compare observed tool effects with the expected business state.")
    tools = [s for s in spans if s["kind"] == "TOOL"]
    for name in case.get("requiredTools", []):
        check("required-tool:" + name, any(s["name"] == name and s["status"] == "ok" for s in tools), "A successful recorded tool call is required.")
    for name in case.get("forbiddenTools", []):
        check("forbidden-tool:" + name, not any(s["name"] == name for s in tools), "This tool must not be invoked.")
    for field, kind, label in (("maxToolCalls", "TOOL", "tool"), ("maxModelCalls", "LLM", "model")):
        if field in case:
            count = sum(s["kind"] == kind for s in spans)
            check(label + "-call-limit", count <= case[field], f"{count} recorded calls; limit {case[field]}.")
    for tool in tools:
        if tool["status"] == "error":
            findings.append(dict(category="tool", title=tool["name"] + " failed", evidence_span_ids=[tool["span_id"]], confidence="observed_failure", fix_brief="Inspect this tool's contract and dependencies. Reproduce the failure, fix it, and rerun incident and regression cases."))
    for item in checks:
        if not item["passed"]:
            findings.append(dict(category="quality" if item["id"] in ("expected-output", "expected-state") else "harness", title=item["id"], evidence_span_ids=[s["span_id"] for s in spans], confidence="hypothesis", fix_brief="Inspect the real application path, tools, state, routing and retries. Change one suspected cause and rerun unchanged incident, regression and holdout cases."))
    status = "error" if error else "unscored" if not checks else "passed" if all(c["passed"] for c in checks) else "failed"
    return dict(id=case["id"], split=case.get("split", "regression"), status=status, checks=checks, findings=findings, spans=spans, **({"error": error} if error else {}))


async def evaluate(bench, *, source_revision: str, context_revision: str, cases: list[dict], run: Callable,
                   observe: Callable | None = None, timeout: float = 30, cancel_event: threading.Event | None = None) -> dict:
    pinned = validate(source_revision, context_revision, cases, timeout)
    if not callable(run) or (observe is not None and not callable(observe)):
        raise ValueError("A real application entry point is required.")
    if bench._closed or bench._evaluation.get() is not None:
        raise ValueError("Use an open client outside another application evaluation.")
    results = []
    for case in pinned:
        if cancel_event and cancel_event.is_set():
            break
        capture = Capture()
        context = EvaluationContext(case["id"], threading.Event(), time.monotonic() + timeout)
        async def execute():
            own = bench._evaluation.set(capture)
            parent = bench._context.set(None)
            try:
                with bench.trace("system-entrypoint", kind="AGENT", input=snapshot(case["input"])) as root:
                    output = snapshot(await invoke(run, snapshot(case["input"]), context))
                    state = snapshot(await invoke(observe, context)) if observe else None
                    root.set_output(output)
                return output, state
            finally:
                bench._context.reset(parent)
                bench._evaluation.reset(own)
        async def cancellation():
            while not (cancel_event and cancel_event.is_set()):
                await asyncio.sleep(0.01)
        task, cancel = asyncio.create_task(execute()), asyncio.create_task(cancellation())
        error, output, state, observed, stopped = None, None, None, False, False
        try:
            done, _ = await asyncio.wait([task, cancel], timeout=timeout, return_when=asyncio.FIRST_COMPLETED)
            if task not in done or context.cancelled or (cancel_event and cancel_event.is_set()):
                stopped = True
                error = "Application timed out or was stopped. No complete score."
            else:
                try:
                    output, state = task.result()
                    observed = observe is not None
                except (Exception, asyncio.CancelledError):
                    error = "Application execution failed. Inspect recorded spans."
        finally:
            context.signal.set()
            cancel.cancel()
            if not task.done():
                task.cancel()
            # Consume eventual exceptions without waiting forever for uncooperative code.
            task.add_done_callback(lambda t: t.exception() if not t.cancelled() else None)
            spans, limited, pending = capture.freeze()
        if limited or pending:
            error = "Application evidence is incomplete. Await all tools and stay within capture limits."
            stopped = stopped or bool(pending)
        result = score(case, output, state, observed, spans, error)
        try:
            result["case_definition"] = bench._safe(case)
            result["output"] = bench._safe(output)
            if observed:
                result["observedState"] = bench._safe(state)
        except Exception:
            result.update(status="error", error="Application evidence could not be redacted.")
        results.append(result)
        if stopped:
            break
    counts = {"passed": 0, "failed": 0, "errors": 0, "unscored": 0}
    for case in results:
        counts["errors" if case["status"] == "error" else case["status"]] += 1
    complete = len(results) == len(pinned) and not counts["errors"] and not counts["unscored"]
    digest = hashlib.sha256(json.dumps(dict(context=context_revision, cases=pinned), sort_keys=True, separators=(",", ":"), ensure_ascii=False).encode()).hexdigest()
    return dict(schema_version=1, execution_mode="application_runtime", evidence_origin="sdk_client_reported",
                environment=bench._environment or "unspecified", source_revision=source_revision, context_revision=context_revision,
                suite_hash=digest, planned_case_count=len(pinned), coverage=dict(instrumentation="explicit_spans", tool_dependencies="application_configured", hosted_validation=False),
                cases=results, summary=dict(status="completed" if complete else "incomplete", score=100 * counts["passed"] / len(pinned) if complete else None, **counts))


async def simulate(bench, *, create_session: Callable, **options) -> dict:
    if not callable(create_session):
        raise ValueError("An application session factory is required.")
    cases = snapshot(options["cases"])
    for case in cases:
        data = case.get("input", {})
        if not isinstance(data, dict) or not isinstance(data.get("turns"), list) or not 1 <= len(data["turns"]) <= 20:
            raise ValueError("A simulation requires 1 to 20 scripted user turns.")
        if "expected_state" not in case and "expectedState" not in case:
            raise ValueError("A simulation requires the expected business state.")
        if "initial_state" in data:
            if "initialState" in data:
                raise ValueError("Use one spelling for initial state.")
            data["initialState"] = data.pop("initial_state")
        if "initialState" not in data:
            raise ValueError("Provide initial_state for a fresh simulation session.")
    observations = {}
    async def run(data, context):
        session = await invoke(create_session, data["initialState"], context)
        try:
            if not all(callable(getattr(session, method, None)) for method in ("turn", "observe", "close")):
                raise ValueError("A session must provide turn, observe and close.")
            reply = None
            for index, message in enumerate(data["turns"]):
                context.raise_if_cancelled()
                turn = EvaluationContext(context.case_id, context.signal, context.deadline, index)
                reply = await invoke(session.turn, message, turn)
            context.raise_if_cancelled()
            observations[context.case_id] = snapshot(await invoke(session.observe))
            return reply
        finally:
            if callable(getattr(session, "close", None)):
                await invoke(session.close)
    def observe(context):
        return observations.pop(context.case_id)
    return await evaluate(bench, **{**options, "cases": cases}, run=run, observe=observe)

File: go/evaluation.go
package bench

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"regexp"
	"strings"
	"sync"
	"time"
)

// Expect distinguishes an explicit JSON null assertion from an omitted assertion.
func Expect(value any) *any { return &value }

type SystemCase struct {
	ID              string   `json:"id"`
	Input           any      `json:"input"`
	Split           string   `json:"split,omitempty"`
	ExpectedOutput  *any     `json:"expectedOutput,omitempty"`
	ExpectedState   *any     `json:"expectedState,omitempty"`
	BusinessOutcome string   `json:"businessOutcome,omitempty"`
	RequiredTools   []string `json:"requiredTools,omitempty"`
	ForbiddenTools  []string `json:"forbiddenTools,omitempty"`
	MaxToolCalls    *int     `json:"maxToolCalls,omitempty"`
	MaxModelCalls   *int     `json:"maxModelCalls,omitempty"`
}

// UnmarshalJSON preserves an explicit null expected outcome while pinning cases.
func (c *SystemCase) UnmarshalJSON(data []byte) error {
	type plain SystemCase
	var value plain
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.UseNumber()
	if err := decoder.Decode(&value); err != nil {
		return err
	}
	var fields map[string]json.RawMessage
	if err := json.Unmarshal(data, &fields); err != nil {
		return err
	}
	for key, target := range map[string]**any{"expectedOutput": &value.ExpectedOutput, "expectedState": &value.ExpectedState} {
		if raw, ok := fields[key]; ok {
			var decoded any
			d := json.NewDecoder(bytes.NewReader(raw))
			d.UseNumber()
			if err := d.Decode(&decoded); err != nil {
				return err
			}
			*target = Expect(decoded)
		}
	}
	*c = SystemCase(value)
	return nil
}

type SystemEvaluationOptions struct {
	SourceRevision, ContextRevision string
	Cases                           []SystemCase
	Run                             func(context.Context, any) (any, error)
	Observe                         func(context.Context, string) (any, error)
	Timeout                         time.Duration
}
type SystemCheck struct {
	ID     string `json:"id"`
	Passed bool   `json:"passed"`
	Reason string `json:"reason"`
}
type RuntimeFinding struct {
	Category        string   `json:"category"`
	Title           string   `json:"title"`
	EvidenceSpanIDs []string `json:"evidence_span_ids"`
	Confidence      string   `json:"confidence"`
	FixBrief        string   `json:"fix_brief"`
}
type SystemCaseResult struct {
	ID             string           `json:"id"`
	Split          string           `json:"split"`
	Status         string           `json:"status"`
	CaseDefinition any              `json:"case_definition,omitempty"`
	Output         any              `json:"output"`
	ObservedState  *any             `json:"observedState,omitempty"`
	Spans          []wireSpan       `json:"spans"`
	Checks         []SystemCheck    `json:"checks"`
	Findings       []RuntimeFinding `json:"findings"`
	Error          string           `json:"error,omitempty"`
}
type SystemSummary struct {
	Status   string   `json:"status"`
	Score    *float64 `json:"score"`
	Passed   int      `json:"passed"`
	Failed   int      `json:"failed"`
	Errors   int      `json:"errors"`
	Unscored int      `json:"unscored"`
}
type SystemEvaluationReport struct {
	SchemaVersion    int                `json:"schema_version"`
	ExecutionMode    string             `json:"execution_mode"`
	EvidenceOrigin   string             `json:"evidence_origin"`
	Environment      string             `json:"environment"`
	SourceRevision   string             `json:"source_revision"`
	ContextRevision  string             `json:"context_revision"`
	SuiteHash        string             `json:"suite_hash"`
	PlannedCaseCount int                `json:"planned_case_count"`
	Coverage         map[string]any     `json:"coverage"`
	Cases            []SystemCaseResult `json:"cases"`
	Summary          SystemSummary      `json:"summary"`
}

// Passed is a strict CI gate: complete coverage, no failed/error/unscored cases.
func (r SystemEvaluationReport) Passed() bool {
	if r.Summary.Status != "completed" || r.PlannedCaseCount == 0 || len(r.Cases) != r.PlannedCaseCount {
		return false
	}
	for _, c := range r.Cases {
		if c.Status != "passed" {
			return false
		}
	}
	return true
}

type evaluationCapture struct {
	mu              sync.Mutex
	active, limited bool
	pending         int
	spans           []wireSpan
}

func (e *evaluationCapture) start() { e.mu.Lock(); defer e.mu.Unlock(); e.pending++ }
func (e *evaluationCapture) finish(row *wireSpan) {
	e.mu.Lock()
	defer e.mu.Unlock()
	e.pending--
	if e.active {
		if row == nil || len(e.spans) >= 100 {
			e.limited = true
		} else {
			e.spans = append(e.spans, *row)
		}
	}
}
func (e *evaluationCapture) freeze() ([]wireSpan, bool, int) {
	e.mu.Lock()
	defer e.mu.Unlock()
	e.active = false
	return append([]wireSpan{}, e.spans...), e.limited, e.pending
}

func snapshot(value any) (any, error) {
	raw, err := json.Marshal(value)
	if err != nil {
		return nil, err
	}
	var result any
	d := json.NewDecoder(bytes.NewReader(raw))
	d.UseNumber()
	err = d.Decode(&result)
	return result, err
}
func jsonEqual(a, b any) bool {
	x, e := json.Marshal(a)
	y, f := json.Marshal(b)
	return e == nil && f == nil && bytes.Equal(x, y)
}
func (c *Client) safeEvidence(value any) (result any, err error) {
	defer func() {
		if recover() != nil {
			err = errors.New("application evidence could not be redacted")
		}
	}()
	return c.safe(value)
}

func validateEvaluation(options SystemEvaluationOptions) ([]SystemCase, error) {
	if !regexp.MustCompile(`^[a-f0-9]{40}$`).MatchString(options.SourceRevision) || len(options.ContextRevision) < 1 || len(options.ContextRevision) > 200 || options.Run == nil {
		return nil, errors.New("pin source and context revisions and provide a real application entry point")
	}
	if len(options.Cases) < 1 || len(options.Cases) > 100 || options.Timeout < 0 || options.Timeout > 300*time.Second {
		return nil, errors.New("use 1 to 100 cases and a timeout up to 300 seconds")
	}
	raw, err := json.Marshal(options.Cases)
	if err != nil {
		return nil, err
	}
	if len(raw) > 500000 {
		return nil, errors.New("case suite exceeds 500 KB")
	}
	var cases []SystemCase
	if err = json.Unmarshal(raw, &cases); err != nil {
		return nil, err
	}
	ids := map[string]bool{}
	for _, item := range cases {
		if item.ID == "" || len(item.ID) > 100 || ids[item.ID] {
			return nil, errors.New("case IDs must be unique and nonempty")
		}
		ids[item.ID] = true
		switch item.Split {
		case "", "capability", "incident", "regression", "holdout":
		default:
			return nil, errors.New("invalid case split")
		}
		if len(item.BusinessOutcome) > 2000 || len(item.RequiredTools)+len(item.ForbiddenTools) > 90 {
			return nil, errors.New("case assertions exceed limits")
		}
		for _, names := range [][]string{item.RequiredTools, item.ForbiddenTools} {
			for _, name := range names {
				if len(name) < 1 || len(name) > 200 {
					return nil, errors.New("invalid tool name")
				}
			}
		}
		for _, limit := range []*int{item.MaxToolCalls, item.MaxModelCalls} {
			if limit != nil && (*limit < 0 || *limit > 10000) {
				return nil, errors.New("call limits must be nonnegative integers up to 10000")
			}
		}
	}
	return cases, nil
}

func scoreCase(item SystemCase, output, state any, observed bool, spans []wireSpan, failure string) SystemCaseResult {
	r := SystemCaseResult{ID: item.ID, Split: item.Split, Spans: spans, Checks: []SystemCheck{}, Findings: []RuntimeFinding{}, Output: output, Error: failure}
	if r.Split == "" {
		r.Split = "regression"
	}
	root := false
	tools, models := 0, 0
	ids := []string{}
	for _, s := range spans {
		ids = append(ids, s.ID)
		if s.Parent == "" && s.Kind == "AGENT" && s.Status == "ok" {
			root = true
		}
		if s.Kind == "TOOL" {
			tools++
		}
		if s.Kind == "LLM" {
			models++
		}
	}
	if !root && r.Error == "" {
		r.Error = "Application root was not recorded. No complete score."
	}
	if item.ExpectedState != nil && !observed && r.Error == "" {
		r.Error = "Application state was not observed. No complete score."
	}
	check := func(id string, passed bool, reason string) {
		r.Checks = append(r.Checks, SystemCheck{id, passed, reason})
	}
	if item.ExpectedOutput != nil {
		check("expected-output", jsonEqual(output, *item.ExpectedOutput), "Compare actual output with the expected outcome.")
	}
	if item.ExpectedState != nil && observed {
		check("expected-state", jsonEqual(state, *item.ExpectedState), "Compare observed tool effects with expected business state.")
	}
	for _, name := range item.RequiredTools {
		found := false
		for _, s := range spans {
			found = found || (s.Kind == "TOOL" && s.Name == name && s.Status == "ok")
		}
		check("required-tool:"+name, found, "A successful recorded tool call is required.")
	}
	for _, name := range item.ForbiddenTools {
		found := false
		for _, s := range spans {
			found = found || (s.Kind == "TOOL" && s.Name == name)
		}
		check("forbidden-tool:"+name, !found, "This tool must not be invoked.")
	}
	if item.MaxToolCalls != nil {
		check("tool-call-limit", tools <= *item.MaxToolCalls, fmt.Sprintf("%d recorded tool calls", tools))
	}
	if item.MaxModelCalls != nil {
		check("model-call-limit", models <= *item.MaxModelCalls, fmt.Sprintf("%d recorded model calls", models))
	}
	for _, s := range spans {
		if s.Kind == "TOOL" && s.Status == "error" {
			r.Findings = append(r.Findings, RuntimeFinding{"tool", s.Name + " failed", []string{s.ID}, "observed_failure", "Inspect this tool's contract and dependencies. Reproduce the failure and rerun incident and regression cases."})
		}
	}
	passed := true
	for _, check := range r.Checks {
		if !check.Passed {
			passed = false
			category := "harness"
			if check.ID == "expected-output" || check.ID == "expected-state" {
				category = "quality"
			}
			r.Findings = append(r.Findings, RuntimeFinding{category, check.ID, ids, "hypothesis", "Inspect the real application path, tools, state, routing and retries. Rerun unchanged incident, regression and holdout cases."})
		}
	}
	switch {
	case r.Error != "":
		r.Status = "error"
	case len(r.Checks) == 0:
		r.Status = "unscored"
	case passed:
		r.Status = "passed"
	default:
		r.Status = "failed"
	}
	return r
}

// EvaluateSystem runs callbacks locally; it never silently publishes a report or calls a judge.
// Callbacks must honor their context. A deadline cannot forcibly terminate a goroutine.
func (c *Client) EvaluateSystem(ctx context.Context, options SystemEvaluationOptions) (SystemEvaluationReport, error) {
	cases, err := validateEvaluation(options)
	if err != nil {
		return SystemEvaluationReport{}, err
	}
	c.mu.Lock()
	closed := c.closed
	c.mu.Unlock()
	parent, _ := ctx.Value(contextKey{c}).(traceContext)
	if closed || parent.evaluation != nil {
		return SystemEvaluationReport{}, errors.New("use an open client outside another application evaluation")
	}
	timeout := options.Timeout
	if timeout == 0 {
		timeout = 30 * time.Second
	}
	pinned, _ := json.Marshal(map[string]any{"context": options.ContextRevision, "cases": cases})
	digest := sha256.Sum256(pinned)
	env := c.options.Environment
	if env == "" {
		env = "unspecified"
	}
	report := SystemEvaluationReport{SchemaVersion: 1, ExecutionMode: "application_runtime", EvidenceOrigin: "sdk_client_reported", Environment: env, SourceRevision: options.SourceRevision, ContextRevision: options.ContextRevision, SuiteHash: hex.EncodeToString(digest[:]), PlannedCaseCount: len(cases), Cases: []SystemCaseResult{}, Coverage: map[string]any{"instrumentation": "explicit_spans", "tool_dependencies": "application_configured", "hosted_validation": false}}
	for _, item := range cases {
		if ctx.Err() != nil {
			break
		}
		capture := &evaluationCapture{active: true}
		runCtx, cancel := context.WithTimeout(ctx, timeout)
		runCtx = context.WithValue(runCtx, contextKey{c}, traceContext{evaluation: capture})
		runCtx = context.WithValue(runCtx, evaluationCaseKey{}, item.ID)
		type outcome struct {
			output, state any
			observed      bool
			failed        bool
		}
		done := make(chan outcome, 1)
		go func(item SystemCase) {
			result := outcome{}
			defer func() {
				if recover() != nil {
					result.failed = true
				}
				done <- result
			}()
			input, e := snapshot(item.Input)
			if e != nil {
				result.failed = true
				return
			}
			inner, root := c.StartSpan(runCtx, SpanInput{Name: "system-entrypoint", Kind: "AGENT", Input: input})
			defer root.End()
			output, e := options.Run(inner, input)
			if e != nil {
				result.failed = true
				return
			}
			result.output, e = snapshot(output)
			if e != nil {
				result.failed = true
				return
			}
			if options.Observe != nil {
				result.state, e = options.Observe(inner, item.ID)
				if e == nil {
					result.state, e = snapshot(result.state)
				}
				result.observed = e == nil
				result.failed = e != nil
			}
			if !result.failed {
				root.SetOutput(output)
			}
		}(item)
		result := outcome{}
		failure := ""
		stopped := false
		select {
		case result = <-done:
			if result.failed {
				failure = "Application execution failed. Inspect recorded spans."
			}
		case <-runCtx.Done():
			stopped = true
			failure = "Application timed out or was stopped. No complete score."
		}
		if runCtx.Err() != nil {
			stopped = true
			failure = "Application timed out or was stopped. No complete score."
		}
		cancel()
		spans, limited, pending := capture.freeze()
		if limited || pending > 0 {
			failure = "Application evidence is incomplete. Await all tools and stay within capture limits."
			stopped = stopped || pending > 0
		}
		row := scoreCase(item, result.output, result.state, result.observed, spans, failure)
		row.CaseDefinition, err = c.safeEvidence(item)
		if err == nil {
			row.Output, err = c.safeEvidence(result.output)
		}
		if err == nil && result.observed {
			var state any
			state, err = c.safeEvidence(result.state)
			row.ObservedState = Expect(state)
		}
		if err != nil {
			row.Status = "error"
			row.Error = "Application evidence could not be redacted."
			row.Output = nil
		}
		report.Cases = append(report.Cases, row)
		if stopped {
			break
		}
	}
	report.Summary.Status = "incomplete"
	for _, item := range report.Cases {
		switch item.Status {
		case "passed":
			report.Summary.Passed++
		case "failed":
			report.Summary.Failed++
		case "error":
			report.Summary.Errors++
		default:
			report.Summary.Unscored++
		}
	}
	if len(report.Cases) == len(cases) && report.Summary.Errors == 0 && report.Summary.Unscored == 0 {
		report.Summary.Status = "completed"
		score := 100 * float64(report.Summary.Passed) / float64(len(cases))
		report.Summary.Score = &score
	}
	return report, nil
}

// PublishSystemEvaluation explicitly saves evidence without running a paid evaluation.
func (c *Client) PublishSystemEvaluation(ctx context.Context, systemID int64, report SystemEvaluationReport) (err error) {
	defer func() {
		if recover() != nil {
			err = errors.New("application report could not be redacted")
		}
	}()
	if systemID < 1 || systemID > 9007199254740991 {
		return errors.New("select a system")
	}
	var value any = report
	if c.options.Redact != nil {
		value = c.options.Redact(value)
	}
	value, err = snapshot(value)
	if err != nil {
		return errors.New("invalid application report")
	}
	body, err := json.Marshal(scrubLimit(value, 0, 300))
	if err != nil {
		return err
	}
	if len(body) > 500000 {
		return errors.New("runtime report exceeds 500 KB; retain it locally or split the suite")
	}
	url := strings.TrimSuffix(c.endpoint, "/api/traces") + fmt.Sprintf("/api/ai-systems/%d/runtime-evaluations", systemID)
	request, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return err
	}
	request.Header.Set("Authorization", "Bearer "+c.options.APIKey)
	request.Header.Set("Content-Type", "application/json")
	response, err := c.http.Do(request)
	if err != nil {
		return errors.New("could not save application results; local results remain available")
	}
	defer response.Body.Close()
	_, _ = io.Copy(io.Discard, io.LimitReader(response.Body, 4096))
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return fmt.Errorf("could not save application results (HTTP %d); local results remain available", response.StatusCode)
	}
	return nil
}
